### PR TITLE
drivers: dma: add dma_emul driver and test coverage

### LINKED ISSUE #57129: drivers: dma: emul driver
**Is your feature request related to a problem? Please describe.**
Many Zephyr drivers incorporate what we call an `emul` driver. Since DMA is so critical to so many types of IP, it would be great to be able to have a software implementation of it to "hook up" or "stitch together" other emulated device drivers.

Covered to some extent by @sjg20 's and @cfriedt's talks at ZDS 2022

Overall, `emul` drivers provide a number of benefits
* prototypical model for Driver API for the specific class of driver
  * physical drivers must mimic this behaviour
* driver-specific DT bindings to allow for customization on a per-instance basis
* drop-in replacement via DT phandle for emulated app regression testing
* **Fast, Functional Testing** for application code (regression testing, prototyping, ..)
  * also includes fast functional sample applications
* private API extensions
  * customizable backend functionality (e.g. "wiring" for `gpio_emul`)
  * test and cover very specific corner cases
  * such corner cases can be very difficult to reproduce in reality
  * coverage in safety-critical and high-reliability systems can be extremely important (think Automotive, Avionics, Infrastructure, ...)
* Can be instantiated on any board, regardless of architecture (`native_posix`, `qemu*`, actual physical boards)
* Can be used for more than just testing (e.g. `gpio_emul_sdl.c`)

**Describe the solution you'd like**
* Probably an RFC to start off with
  * Outline DT bindings, testsuite integration, strategies
  * Buy-in from DMA subsystem maintainer @teburd
  * Suggestions / feedback from other `emul` driver authors

**Describe alternatives you've considered**
Not having a DMA `emul` driver (which is the case today). Not that anything is wrong with the DMA subsystem today, but we want **Fast Functional Testing**!

**Additional context**
May coincide with existing activities by @jgl-meta

### PR DESCRIPTION
**Pull request by cfriedt**
4 commits (please see each commit message for details)

* dts: bindings: dma: add bindings for an emulated dma controller
* drivers: dma: add emulated dma driver
* boards: posix: add dma support to native_posix boards 
* tests: drivers: dma: test dma_emul driver on native_posix boards

It's pretty basic right now, but the intention is to support interfaces other than simple `memcpy()` via a backend API. With that, we should be able to use e.g. TCP or UDP sockets, shared memory, etc.

Additionally, some drivers (such as PCIe) often depend on DMA to be present, so this has the potential to open up ZTest to areas that have traditionally not been very testable. This could, for example, pair pretty well with the recent NVMe API.

Fixes https://github.com/zephyrproject-rtos/zephyr/pull/57129

### CODE REVIEW

**Comment by cfriedt** [comment]
* rebased to fix merge conflict
* added license to `drivers/dma/dma_emul.c`

**Comment by cfriedt** [comment]
* use `CONFIG_DMA_LOG_LEVEL`

**Comment by cfriedt** [comment]
I anticipate a few rounds to figure out any loose ends with the implementation. Additionally, I may have found a couple of gaps here and there with the current DMA API and testsuite that could be closed.

**Comment by cfriedt** [comment]
* disabled `native_posix*` for `scatter_gather` test
* added missing `dma-channels = <2>` to `native_posix.overlay`

**Comment by cfriedt** [comment]
unrelated error is unrelated
https://github.com/zephyrproject-rtos/zephyr/issues/58376

File: dts/bindings/dma/zephyr,dma-emul.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  stack-size:
+    type: int
+    required: true
+    description: >
+      Stack size (in bytes) for the instance-specific work_q thread.
+
+  priority:
+    type: int
+    description: >
+      Priorityfor the instance-specific work_q thread.
```
Comment (ycsin) [inline dts/bindings/dma/zephyr,dma-emul.yaml:20]:
```suggestion
      Priority for the instance-specific work_q thread.
```

**Comment by cfriedt** [comment]
* increased `dma-channels` to 4 for `scatter_gather` tests
* additional checks in `dma_emul_config_valid()` to verify availability of slots for chained transfers

File: dts/bindings/dma/zephyr,dma-emul.yaml (hunk 1 above)
Comment (cfriedt) [reply dts/bindings/dma/zephyr,dma-emul.yaml]: Fixed!

**Comment by cfriedt** [comment]
* corrected typo in `zephyr,dma-emul.yaml`

**Comment by cfriedt** [comment]
I'll put together an RFC as well. Might be good to schedule this for the arch meeting @carlescufi, just because it would be nice to document the model, expected behaviour, and fill any gaps in testing that are not covered.

File: drivers/dma/dma_emul.c [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+			memcpy(&block,
+			       &config->block[channel * config->num_requests +
+					      xfer_config.dma_slot + i],
+			       sizeof(block));
+			k_spin_unlock(&data->lock, key);
+
+			/* transfer data in bursts */
+			for (bytes = MIN(block.block_size, xfer_config.dest_burst_length);
+			     bytes > 0; block.block_size -= bytes, block.source_address += bytes,
+			    block.dest_address += bytes, remaining -= bytes,
```
Comment (ycsin) [inline drivers/dma/dma_emul.c:243]: `remaining` not init before used:

```
/__w/zephyr/zephyr/drivers/dma/dma_emul.c:243:37: error: variable 'remaining' is uninitialized when used here [-Werror,-Wuninitialized]
                            block.dest_address += bytes, remaining -= bytes,
                                                         ^~~~~~~~~
/__w/zephyr/zephyr/drivers/dma/dma_emul.c:201:18: note: initialize the variable 'remaining' to silence this warning
        size_t remaining;
                        ^
                         = 0
1 error generated.
```

**Comment by cfriedt** [comment]
* removed unused `remaining` variable

File: drivers/dma/dma_emul.c (hunk 2 above)
Comment (cfriedt) [reply drivers/dma/dma_emul.c]: Fixed!

**Review by teburd — COMMENTED** [review]
In looking this over I think its pretty reasonable. It's in effect a software DMA controller. I'd love to better understand why in particular an emulated DMA is needed as almost all of Zephyr hides DMA under peripheral API layers and lets DMA do whatever is needed on that particular platform. Meaning the API is in effect a union of all DMA behaviors available among hardware vendors. There's absolutely some overlap, but many areas where hardware specifics are in play.

E.g. how does DMA interact with a SPI peripheral? This varies *wildly* from what I've seen. But since it seems useful to at least one person I think that's enough for me.

I'll happily +1 so long as a few things are documented. The caveat that this emulated DMA can and should be extended with various Kconfig/dt behavioral options to match this varied behavior of real hardware as needed is documented somewhere. That the intent is not to be a definitive on how DMA drivers should be implemented but as a mock like emulated device for use in testing. A doc comment in the driver somewhere is enough for me in that regard.

**Comment by cfriedt** [comment]
@teburd - thanks for taking another look.

> In looking this over I think its pretty reasonable. It's in effect a software DMA controller.

Yes, exactly. There will be some flexibility in terms of backends and some customizable behaviour to an extent too.

> how does DMA interact with a SPI peripheral? This varies wildly from what I've seen.

That's a good question, and to be honest, not one that I've considered yet, but definitely agree that it varies wildly. 

It might be easiest to tackle specific peripherals / interconnects on a case-by-case basis. SPI is on the radar, but my MVP in the next while will be PCIe DMA, and I would be looking to emulate something close to the "edu" device in Qemu. It was touched upon in Zephyr [here](https://github.com/zephyrproject-rtos/zephyr/pull/48865), with additional info [here](https://github.com/qemu/qemu/blob/master/docs/specs/edu.txt) and [here](https://github.com/qemu/qemu/blob/master/hw/misc/edu.c).

The "edu" device mainly has a host-side specification, so there is some flexibility for device-side (where Zephyr would mainly be "running").

It will be a fair bit of work to get there, but we will have additional collaboration within the community (I added a bunch of AntMicro folks to review very deliberately 😁, cc @mgielda)

**Comment by cfriedt** [comment]
I'll rebase this just to make sure it gets back into a mergeable state. We may be able to begin working on it in the near term again.

**Comment by cfriedt** [comment]
* rebased

**Comment by MaureenHelm** [comment]
Needs a rebase

**Comment by cfriedt** [comment]
* rebased (cc @PiotrZierhoffer)

Screenshot 2023-11-14 at 11 27 06 AM

**Comment by cfriedt** [comment]
rebased again

**Review by teburd — DISMISSED** [review]

**Comment by cfriedt** [comment]
* fixed merge conflicts

@teburd - can you please re-ack?

**Review by teburd — DISMISSED** [review]

**Comment by cfriedt** [comment]
* renamed `reserved` to `_reserved`

**Review by ycsin — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Pull request merged**